src/charm.py

```diff
@@ -108,12 +110,15 @@ def to_status(tpl: Tuple[str, str]) -> StatusBase:
         MetricsEndpointProvider,
     ],
 )
+@log_charm(logging_endpoints="logging_endpoints", server_cert="server_ca_cert_path")
```

Comment:
In [loki docs](https://grafana.com/docs/loki/latest/send-data/), they list a few loki clients in python, and two of them work directly with the logging module:

- https://pypi.org/project/python-logging-loki/
- https://pypi.org/project/nextlog/

Could we simplify by a call to `logger.addHandler(...)` from the class's `__init__`?

Reply:
I tried out python-logging-loki and
1) it wasn't working out of the box, also it missed tls integration capability
2) it was using requests

so I ended up copying most of the code from it and fixing what was broken/missing. Instead of adding an external dependency I'd rather vendor what we need in the charm lib, since it's not much code anyway.

Reply:
Secondly, which class are you referring to? the charm? 
Just like in charm tracing, we want to set this up before the charm's init runs, so that we can also capture more ops logs and avoid any races in init

Reply:
resolved in conversations: moved lokihandler and related code to cosl in https://github.com/canonical/cos-lib/pull/26